Sources/Jobs/JobQueueHandler.swift

```diff
@@ -68,54 +67,66 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Sendable {
     func runJob(_ queuedJob: QueuedJob<Queue.JobID>) async throws {
         var logger = logger
         let startTime = DispatchTime.now().uptimeNanoseconds
-
-        Meter(label: self.meterLabel, dimensions: [("status", JobStatus.queued.rawValue)]).decrement()
-        Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
-        defer {
-            Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
-        }
         logger[metadataKey: "JobID"] = .stringConvertible(queuedJob.id)
         let job: any JobInstanceProtocol
         do {
             job = try self.jobRegistry.decode(queuedJob.jobBuffer)
         } catch let error as JobQueueError where error == .unrecognisedJobId {
             logger.debug("Failed to find Job with ID while decoding")
             try await self.queue.failed(jobId: queuedJob.id, error: error)
+            Meter(label: JobMetricsHelper.discardedMeter, dimensions: [
+                ("reason", "INVALID_JOB_ID"),
+                ("jobID", queuedJob.id.description),
+            ]).increment()
             return
         } catch {
             logger.debug("Job failed to decode")
             try await self.queue.failed(jobId: queuedJob.id, error: JobQueueError.decodeJobFailed)
+            Meter(label: JobMetricsHelper.discardedMeter, dimensions: [
+                ("reason", "DECODE_FAILED"),
+                ("jobID", queuedJob.id.description),
+            ]).increment()
             return
         }
         logger[metadataKey: "JobName"] = .string(job.name)
 
         // Calculate wait time from queued to processing
         let jobQueuedDuration = Date.now.timeIntervalSince(job.queuedAt)
         Timer(
-            label: "\(self.metricsLabel).queued.duration",
+            label: "\(JobMetricsHelper.metricsLabel).queued.duration",
             dimensions: [("name", job.name)],
             preferredDisplayUnit: .seconds
         ).recordSeconds(jobQueuedDuration)
 
+        // Decrement the current job by 1
+        Meter(label: JobMetricsHelper.meterLabel, dimensions: [
+            ("status", JobMetricsHelper.JobStatus.queued.rawValue),
+            ("name", job.name),
+        ]).decrement()
```

Comment:
I think this call should go at the top of the function, as the job has been removed from the queue by then and if the code fails to decode it doesnt decrement the metric.

Reply:
I will move this call at the top. I still have one question about ways of pushing jobs in the queue. [There's this push func](https://github.com/hummingbird-project/swift-jobs/blob/main/Sources/Jobs/JobQueue.swift#L45) and job parameters [push](https://github.com/hummingbird-project/swift-jobs/blob/main/Sources/Jobs/JobParameters.swift#L28) which calls JobQueue [push](https://github.com/hummingbird-project/swift-jobs/blob/main/Sources/Jobs/JobParameters.swift#L39) I think we are missing queue inc between these funcs. JobScheduler [uses](https://github.com/hummingbird-project/swift-jobs/blob/main/Sources/Jobs/JobParameters.swift#L39). We might see negative count depending on how a job was pushed onto a queue.

Reply:
The user shouldn't be calling push on the driver, they need to do it through the queue. I'm not sure how we can be sure people don't do this though. In theory they could call any of the function on the driver and break stuff

Reply:
I see! Should the scheduler use this push [func](https://github.com/hummingbird-project/swift-jobs/blob/main/Sources/Jobs/JobQueue.swift#L45C36-L45C40) instead? Otherwise, there needs to be an Meter increment [here](https://github.com/hummingbird-project/swift-jobs/blob/main/Sources/Jobs/Scheduler/JobSchedule.swift#L187)

Reply:
The scheduler probably should but I think there was an issue getting that working. We might need to move the JobQueueDriver extension onto JobQueue if I remember right. Let's look into that in a separate PR. I'd really like to land this one and get a release out with the new swift-jobs-postgres repo as well.